Source/WebCore/rendering/RenderText.cpp

```diff
@@ -1329,6 +1329,55 @@ UChar RenderText::previousCharacter() const
     return previousString[previousString.length() - 1];
 }
 
+// https://www.w3.org/TR/css-text-3/#small-kana
+static String convertToFullSizeKana(const String& string)
+{
+    static const uint16_t smallKanas[] = {
+        // ぁ   ぃ      ぅ      ぇ      ぉ      っ      ゃ      ゅ      ょ
+        0x3041, 0x3043, 0x3045, 0x3047, 0x3049, 0x3063, 0x3083, 0x3085, 0x3087,
+        // ゎ   ゕ      ゖ
+        0x308E, 0x3095, 0x3096,
+        // ァ   ィ      ゥ      ェ      ォ      ッ      ャ      ュ      ョ
+        0x30A1, 0x30A3, 0x30A5, 0x30A7, 0x30A9, 0x30C3, 0x30E3, 0x30E5, 0x30E7,
+        // ヮ   ヵ      ヶ      ㇰ      ㇱ      ㇲ      ㇳ      ㇴ      ㇵ
+        0x30EE, 0x30F5, 0x30F6, 0x31F0, 0x31F1, 0x31F2, 0x31F3, 0x31F4, 0x31F5,
+        // ㇶ   ㇷ      ㇸ      ㇹ      ㇺ      ㇻ      ㇼ      ㇽ      ㇾ
+        0x31F6, 0x31F7, 0x31F8, 0x31F9, 0x31FA, 0x31FB, 0x31FC, 0x31FD, 0x31FE,
+        // ㇿ
+        0x31FF,
+        // ｧ    ｨ       ｩ       ｪ       ｫ       ｬ       ｭ       ｮ       ｯ
+        0xFF67, 0xFF68, 0xFF69, 0xFF6A, 0xFF6B, 0xFF6C, 0xFF6D, 0xFF6E, 0xFF6F
+    };
+    static const uint16_t fullSizeKanas[] = {
+        // あ   い      う      え      お      つ      や      ゆ      よ
+        0x3042, 0x3044, 0x3046, 0x3048, 0x304A, 0x3064, 0x3084, 0x3086, 0x3088,
+        // わ   か      け
+        0x308F, 0x304B, 0x3051,
+        // ア   イ      ウ      エ      オ      ツ      ヤ      ユ      ヨ
+        0x30A2, 0x30A4, 0x30A6, 0x30A8, 0x30AA, 0x30C4, 0x30E4, 0x30E6, 0x30E8,
+        // ワ   カ      ケ      ク      シ      ス      ト      ヌ      ハ
+        0x30EF, 0x30AB, 0x30B1, 0x30AF, 0x30B7, 0x30B9, 0x30C8, 0x30CC, 0x30CF,
+        // ヒ   フ      ヘ      ホ      ム      ラ      リ      ル      レ
+        0x30D2, 0x30D5, 0x30D8, 0x30DB, 0x30E0, 0x30E9, 0x30EA, 0x30EB, 0x30EC,
+        // ロ
+        0x30ED,
+        // ｱ    ｲ       ｳ       ｴ       ｵ       ﾔ       ﾕ       ﾖ        ﾂ
+        0xFF71, 0xFF72, 0xFF73, 0xFF74, 0xFF75, 0xFF94, 0xFF95, 0xFF96, 0xFF82
+    };
+
+    StringBuilder converted;
```

Comment:
I know nothing about this code/feature but isn't there a need for a fast path to avoid re-creating a brand new String in the case where no kana is found?

Or is it just not worth optimizing for in this case?

Reply:
This feature is really niche, I don't think we should look for any optimisation there.